clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -2149,6 +2149,14 @@ public void wakeup() {
         this.client.wakeup();
     }
 
+    /**
+     * Checks if the consumer is connected to Kafka server i.e. broker.
+     */
+    @Override
+    public boolean isConnected() {
+        return !coordinator.coordinatorUnknown();
```

Comment:
I am not familiar with the details of the `KafkaConsumer` code base.

\cc @hachikuji Is this a reasonable approach to figure out if a client lost connectivity to the Kafka cluster?

Reply:
Well, I added this method because in the documentation, it stated the following:
```
    /**
     * Check if we know who the coordinator is and we have an active connection
     * @return true if the coordinator is unknown
     */
    public boolean coordinatorUnknown() {
        return checkAndGetCoordinator() == null;
    }
```
It checks for a connection, and if coordinator is null, I think it means disconnected.

Reply:
A few questions comments:

1. The KIP doesn't have any rejected alternatives. Have you considered any? If you look at the state machine, other transitions are detected through the ConsumerRebalanceListener (e.g. partition revocation and partition assignment). Should we not try to reuse that mechanism instead of introducing a new API? KIP-429 adds a new API `onPartitionsLost`. Could that be used? Another option that comes to mind is exposing a metric in the consumer for the last time since the last heartbeat was successfully received. That would be generally useful and might solve this problem. 

2. To be honest, I am not sure you want to add DISCONNECTED to the state machine since it is most often transient. What would you expect an application to do in this state? If this is just about monitoring, then metrics are the usual weapon of choice.

3. Coordinator liveness is a necessary condition for the application to make progress, but it is not sufficient. Would you want to take into account other factors as well, such as the liveness of partition leaders?

Reply:
Thanks for commenting!

So I did some thinking, and from my understanding of the issue, what we are trying to pursue here is to find a way to tell whether or not the _server_ is alive, not the _consumer_. The definition of the DISCONNECTED state basically means that the consumer thread is perfectly healthy, but is unable to connect to a Kafka node for some reason like a server crash. ConsumerRebalanceListener is reliant on the fact that the server is alive to detect partition revocation and assignment. If we are unable to connect to broker, ConsumerRebalanceListener is useless (i.e. there is no nodes to send or receive requests from). 

So in effect, this change in streams also means that we need to add some public API to KafkaConsumer as well as Kafka Streams. I like your metrics option, since it is cleaner to implement (adding a DISCONNECTED to the state machine would mean you would have to deal with a variety of transitions to other states). 

Looking through the comments in this issue's JIRA, I think that ```coordinatorUnknown()``` by itself might not provide enough information. @guozhangwang mentioned in this issue's JIRA that we should expose connection metrics available in ConsumerNetworkClient's ```ClusterConnectionStates``` which indicates which nodes are connected and disconnected 
as stated [here:.](https://issues.apache.org/jira/browse/KAFKA-6520?focusedCommentId=16544335&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16544335).

Reply:
Sorry for getting late on this, I've some some comments on the KIP mailing thread itself on both the high-level motivation and also lower level implementation to achieve the motivation, please let me know if it makes sense.

Reply:
Alright, better chipping in now than never. :)

Reply:
cc @guozhangwang You mentioned in the mailing thread that we should consider a server not to be alive when heartbeat rate drops to zero. So I'm curious when the heartbeat rate is actuallly considered zero. When I thought about it, it seemed to me if we send multiple Heartbeat requests, but never get a response, say within session.timeout.ms, then it can be considered a zero heartbeat rate. How does that sound to you?

Reply:
@mjsax @hachikuji Guozhang mentioned some possible choices on the JIRA, your thoughts on this?